core/src/main/java/io/grpc/Metadata.java

```diff
@@ -396,8 +411,37 @@ private String toStringInternal() {
     private final String name;
     private final byte[] asciiName;
 
+    private static Set<Character> generateValidTChars() {
+      ImmutableSet.Builder<Character> valid = ImmutableSet.builder();
+      valid.add('-');
+      valid.add('_');
+      for (char c = '0'; c <= '9'; c++) {
+        valid.add(c);
+      }
+      for (char c = 'a'; c <= 'z'; c++) {
+        valid.add(c);
+      }
+      return valid.build();
+    }
+
+    private static String validateAndNormalizeName(String n) {
+      checkNotNull(n);
+      char[] token = n.toCharArray();
```

Comment:
I would agree, except that I would generally encourage string-based `toLowerCase` for Unicode/UTF-16 reasons and string-based `toLowerCase` can avoid creating two arrays if it is unnecessary. The Unicode part isn't really all that important here, but most developers wouldn't be able to recognize when it is and isn't important, so I would suggest they always use the string-based form.